BDArmory/Guidances/MissileGuidance.cs

```diff
@@ -268,8 +268,17 @@ public static Vector3 GetAirToAirFireSolution(MissileBase missile, Vessel target
             leadTime = targetDistance / (float)(targetVessel.Velocity() - simMissileVel).magnitude;
             leadTime = Mathf.Clamp(leadTime, 0f, 8f);
             */
+            float vel = (float)missile.vessel.Velocity().magnitude;
+            float VelOpt = (launcher != null ? launcher.optimumAirspeed : 1500) + vel;
             float accel = launcher.thrust / missile.part.mass;
-            leadTime = ((float)Math.Sqrt(accel * (accel + (8 * targetDistance))) - accel) / (2 * accel); //inverse equation for accel to find time from known force and vel
+            float deltaVel = (float)targetVessel.Velocity().magnitude - vel;
+            float DeltaOptvel = (float)targetVessel.Velocity().magnitude - VelOpt;
+            float T = (VelOpt - vel) / accel; //time to optimal airspeed
+            float D = deltaVel * T + 1 / 2 * accel * (T * T); //relative distance to optimum airspeed        
+            float d = targetDistance; 
+            leadTime = (-deltaVel - BDAMath.Sqrt((deltaVel * deltaVel) + 2 * accel * d)) / accel;
```

Comment:
Actually, thinking about this a bit more, `d` should be the distance to the intercept point, not the distance to the target. It may need a couple of iterations of the last few lines updating `d`, starting with `d = targetDistance` and recalculating the distance on where the target will be after that time.

Reply:
Also, the `deltaVel` calculations should be done with `Vector3` then projected onto the forward direction of the missile.

Reply:
Hmm. At that point, we're getting pretty close to basically approaching the commented out trajectorySim code below it from a different angle...
I wonder if we're overthinking this, and could just instead do some thing like this:
```
Vector3 relPosition = targetPosition - missile.transform.position;
Vector3 relAcceleration = targetVessel.acceleration - missile.MissileReferenceTransform.forward * accel;
leadTime = AIUtils.ClosestTimeToCPA(relPosition, deltaVel, relAcceleration, T); //missile accelerating, T is greater than our max look time of 8s
if (T < 8)//missile has reached max speed, and is now cruising; sim positions ahead based on T and run CPA from there
{
    relPosition = (targetPosition + (targetVessel.Velocity() * (8 - T))) - (missile.transform.position + missile.MissileReferenceTransform.forward * (accel * (8 - T)));
    relAcceleration = targetVessel.acceleration; // - missile.MissileReferenceTransform.forward * 0; assume missile is holding steady velocity at optimumAirspeed
    leadTime = AIUtils.ClosestTimeToCPA(relPosition, DeltaOptvel, relAcceleration, 8-T) + T;
}
```

Reply:
At that point we're basically coming close to approaching the commented trajectorySim code form a different angle...
I wonder if we're overthinking things, and could instead just use CPA and do something like this:
```
Vector3 relPosition = targetPosition - missile.transform.position;
Vector3 relAcceleration = targetVessel.acceleration - missile.MissileReferenceTransform.forward * accel;
leadTime = AIUtils.ClosestTimeToCPA(relPosition, deltaVel, relAcceleration, T); //missile accelerating, T is greater than our max look time of 8s
if (T < 8)//missile has reached max speed, and is now cruising; sim positions ahead based on T and run CPA from there
{
    relPosition = (targetPosition + (targetVessel.Velocity() * (8 - T))) - (missile.transform.position + missile.MissileReferenceTransform.forward * (accel * (8 - T)));
    relAcceleration = targetVessel.acceleration; // - missile.MissileReferenceTransform.forward * 0; assume missile is holding steady velocity at optimumAirspeed
    leadTime = AIUtils.ClosestTimeToCPA(relPosition, DeltaOptvel, relAcceleration, 8-T) + T;
}
```

Reply:
The difference between bullets/rockets and missiles is that the missile can steer itself in the direction of the target, which is why the relative position, velocity and acceleration should be projected onto the direction from the launcher to the target (intercept point) and why the equation prior to reaching optimum velocity simplifies to a quadratic. `AIUtils.ClosestTimeToCPA` is assuming that the missile and target follow trajectories with constant acceleration, which the missile doesn't due to being able to steer. If you use the projected relative values in `AIUtils.ClosestTimeToCPA`, then it ought to reduce to the equations above (I think).
The gun aiming code in ModuleWeapon.cs does up to 10 iterations (though I think it usually stops after 2 or 3) to refine the aiming direction, but here you're only interested in the lead time, which shouldn't vary much after the second iteration for `d`.

Reply:
If you're dumb-firing missiles so that they're not steering, then it would make sense to use `AIUtils.ClosestTimeToCPA` (with the missile's acceleration when it's boosting and without it afterwards).

Reply:
This would be for dumbfired missiles, yes, since guided ones are able to do their own course correction and the PilotAI merely needs to point them in the right initial direction (which is what the original code was doing, broadly speaking)

Reply:
Ah ok, that changes things then.
Yeah, using `AIUtils.ClosestTimeToCPA` with `maxTime = T` and including the missile's acceleration should be fine if `leadTime < T`. If not, then update the missile and target's positions with `AIUtils.PredictPosition` and use `AIUtils.ClosestTimeToCPA` again without the missile's acceleration.